### PR TITLE
resolvectl: support to modify interfaces managed by networkd

### LINKED ISSUE #9808: Cannot change DNS servers with resolvectl for networkd-managed interfaces
**Is your feature request related to a problem? Please describe.**
I'd like to be able to change the DNS servers (and in general, any resolved setting) at runtime for a networkd-managed interface.

**Describe the solution you'd like**
``resolvectl dns [if] [....] `` should be able to change dns servers also for networkd-managed interfaces.
I don't see a practical reason as of why this should be a limit, and it's especially useful during debugging, since DHCP interfaces basically override any possible runtime configuration for resolved, which is quite annoying.

**Describe alternatives you've considered**
The only work-around I've found is to un-manage the interface and/or by changing the .network configuration entirely, which is quite annoying given "resolvectl" would be available for that purpose.

### PR DESCRIPTION
**Pull request by yuwata**
Closes #9808.

### CODE REVIEW

**Comment by yuwata** [comment]
Rebased. PTAL.

**Comment by yuwata** [comment]
@mrc0mmand Is it possible to enable and start `systemd-resolved.service` on CentOS CI system? I'd like to add tests for networkd vs resolved.

**Comment by mrc0mmand** [comment]
> @mrc0mmand Is it possible to enable and start `systemd-resolved.service` on CentOS CI system? I'd like to add tests for networkd vs resolved.

It shouldn't be a problem, hopefully. Feel free to add it to the `test-network` (or other) test suite to see how well it behaves :-)

**Comment by yuwata** [comment]
@mrc0mmand Currently, `resolved` seems not running on CentOS CI systems.
```
test_state_file (__main__.NetworkdStateFileTests) ... [0;1;31mFailed to set DNS configuration: Unit dbus-org.freedesktop.resolve1.service not found.[0m

207 dummy98 ether routable configured
expected failure
```

**Comment by mrc0mmand** [comment]
Could you just manually start it like you do with `systemd-networkd.service` in the `test-network` test suite? Or is it going to break something?

**Comment by yuwata** [comment]
OK. I will update so.

**Comment by yuwata** [comment]
Tests for the new features are added, and all CIs are green now. PTAL.

**Comment by yuwata** [comment]
Rebased. PTAL.

**Comment by yuwata** [comment]
Rebased again.

**Comment by yuwata** [comment]
Rebased. PTAL.

**Comment by yuwata** [comment]
@poettering Could you review this?

File: src/network/networkd-manager-bus.c [hunk 1]
```diff
@@ -3,17 +3,57 @@
...
+                return r;
+
+        r = sd_bus_message_open_container(reply, 'a', "(iso)");
+        if (r < 0)
+                return r;
+
+        HASHMAP_FOREACH(link, manager->links, i) {
+                _cleanup_free_ char *path = NULL;
+
+                path = link_bus_path(link);
```
Comment (poettering) [inline src/network/networkd-manager-bus.c:33]: should fail on oom?

File: src/timedate/timedatectl.c [hunk 2]
```diff
@@ -700,6 +700,111 @@ static int show_timesync(int argc, char **argv, void *userdata) {
...
+                        return bus_log_create_error(r);
+                if (r == 0)
+                        return log_error("Failed to find interface '%s'", str);
+
+                if (streq_ptr(n, str)) {
+                        *ret = i;
+                        return 0;;
+                }
+        }
+}
```
Comment (poettering) [inline src/timedate/timedatectl.c:735]: hmm, why not add GetLinkByIndex() as dbus call for this?

**Comment by yuwata** [comment]
@poettering Thank you for the review.

In the revised version, I've tried to use polkit, but it does not work... `CheckAuthorization()` method in `bus_verify_polkit_async()` returns the following error: `Only trusted callers (e.g. uid 0 or an action owner) can use CheckAuthorization() for subjects belonging to other identities`. I picked up the polkit related logics from `timedated`, but `timedated` is running with `root`. How to use polkit from non-root daemons?

Detailed logs:
```
Jul 12 14:15:56 systemd-networkd[29724]: Got message type=method_call sender=:1.331 destination=org.freedesktop.network1 path=/org/freedesktop/network1 interface=org.freedesktop.network1.Manager member=GetLinkByName cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Jul 12 14:15:56 systemd-networkd[29724]: Sent message type=method_return sender=n/a destination=:1.331 path=n/a interface=n/a member=n/a cookie=19 reply_cookie=2 signature=io error-name=n/a error-message=n/a
Jul 12 14:15:56 polkitd[24693]: Registered Authentication Agent for unix-process:29750:1266588 (system bus name :1.332 [/usr/bin/pkttyagent --notify-fd 5 --fallback], object path /org/freedesktop/PolicyKit1/AuthenticationAgent, locale en_US.UTF-8)
Jul 12 14:15:56 systemd-networkd[29724]: Got message type=method_call sender=:1.331 destination=org.freedesktop.network1 path=/org/freedesktop/network1 interface=org.freedesktop.network1.Manager member=RevertLinkNTP cookie=3 reply_cookie=0 signature=i error-name=n/a error-message=n/a
Jul 12 14:15:56 systemd-networkd[29724]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=GetConnectionUnixUser cookie=20 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Jul 12 14:15:56 systemd-networkd[29724]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.330 path=n/a interface=n/a member=n/a cookie=4294967295 reply_cookie=20 signature=u error-name=n/a error-message=n/a
Jul 12 14:15:56 systemd-networkd[29724]: Sent message type=method_call sender=n/a destination=org.freedesktop.PolicyKit1 path=/org/freedesktop/PolicyKit1/Authority interface=org.freedesktop.PolicyKit1.Authority member=CheckAuthorization cookie=21 reply_cookie=0 signature=(sa{sv})sa{ss}us error-name=n/a error-message=n/a
Jul 12 14:15:56 systemd-networkd[29724]: Got message type=error sender=:1.313 destination=:1.330 path=n/a interface=n/a member=n/a cookie=44 reply_cookie=21 signature=s error-name=org.freedesktop.PolicyKit1.Error.NotAuthorized error-message=Only trusted callers (e.g. uid 0 or an action owner) can use CheckAuthorization() for subjects belonging to other identities
Jul 12 14:15:56 systemd-networkd[29724]: Sent message type=error sender=n/a destination=:1.331 path=n/a interface=n/a member=n/a cookie=22 reply_cookie=3 signature=s error-name=org.freedesktop.PolicyKit1.Error.NotAuthorized error-message=Only trusted callers (e.g. uid 0 or an action owner) can use CheckAuthorization() for subjects belonging to other identities
Jul 12 14:15:56 polkitd[24693]: Unregistered Authentication Agent for unix-process:29750:1266588 (system bus name :1.332, object path /org/freedesktop/PolicyKit1/AuthenticationAgent, locale en_US.UTF-8) (disconnected from bus)
Jul 12 14:15:56 systemd-networkd[29724]: Failed to process message type=method_call sender=:1.331 destination=org.freedesktop.network1 path=/org/freedesktop/network1 interface=org.freedesktop.network1.Manager member=RevertLinkNTP cookie=3 reply_cookie=0 signature=i error-name=n/a error-message=n/a: Only trusted callers (e.g. uid 0 or an action owner) can use CheckAuthorization() for subjects belonging to other identities
```

File: src/network/networkd-manager-bus.c (hunk 1 above)
Comment (yuwata) [reply src/network/networkd-manager-bus.c:33]: Yes, fixed.

File: src/timedate/timedatectl.c (hunk 2 above)
Comment (yuwata) [reply src/timedate/timedatectl.c:735]: Implemented.

**Comment by poettering** [comment]
Urks I have no idea, there must be a scheme... Maybe we have to patch Polkit to permit system users to use it too? at least some of them?

**Comment by yuwata** [comment]
I found that setting annotation in .policy file is sufficient. PTAL.

File: src/network/networkd-manager-bus.c [hunk 3]
```diff
@@ -3,17 +3,59 @@
...
+                if (!path)
+                        return -ENOMEM;
+
+                r = sd_bus_message_append(
+                        reply, "(iso)",
+                        link->ifindex,
+                        strempty(link->ifname),
```
Comment (poettering) [inline src/network/networkd-manager-bus.c]: you can drop the strempty() btw, sd_bus_message_append() does that anyway for NULL strings

File: src/network/networkd-manager-bus.c [hunk 4]
```diff
@@ -47,6 +48,75 @@ static int method_list_links(sd_bus_message *message, void *userdata, sd_bus_err
...
+
+        r = sd_bus_message_read(message, "s", &name);
+        if (r < 0)
+                return r;
+
+        HASHMAP_FOREACH(link, manager->links, i)
+                if (streq_ptr(link->ifname, name)) {
+                        found = true;
+                        break;
+                }
```
Comment (poettering) [inline src/network/networkd-manager-bus.c]: hmm, i'd just translate the name via if_nametoindex() and then look it up by index. Sounds faster to me if we have a myriad of devices

**Comment by yuwata** [comment]
@poettering Thank you for the review. I force-pushed a revised version, which addresses the above two comments. PTAL.

**Pull request merged**